### PR TITLE
Tealv5 transaction ops, TxReceipt in runtime

### PR DESCRIPTION
**Pull request by ratik21**
## Proposed Changes

+ added opcodes: `Txnas, Gtxnas, Gtxnsas, Args, Log`. Added tests.
+ Update all transaction functions (eg. `executeTx`, `addAsset`, `addApp` ..etc) to return a transaction receipt. 

NOTE: this introduces a breaking change for `.addAsset` & `.addApp`

## Potential followups
[DONE] integration test for `log` op

### CODE REVIEW

File: packages/runtime/src/runtime.ts [hunk 1]
```diff
@@ -60,6 +63,14 @@ export class Runtime {
     this.timestamp = 1;
   }
 
+  /**
+   * Returns transaction receipt for a particular transaction
+   * @param txID transaction ID
+   */
+  getTransactionInfo (txID: string): TxReceipt | undefined {
```
Comment (robert-zaremba) [inline packages/runtime/src/runtime.ts:70]:
```suggestion
  getTxReceipt (txID: string): TxReceipt | undefined {
```

File: packages/runtime/src/runtime.ts [hunk 2]
```diff
@@ -336,32 +347,38 @@ export class Runtime {
         : asaDef.metadataHash,
       mockSuggestedParams(flags, this.round)
     );
+
+    if (this.ctx.tx === undefined || this.ctx.tx.type !== "acfg") { // could already be defined (if used as a txGroup in this.executeTx())
```
Comment (robert-zaremba) [inline packages/runtime/src/runtime.ts:351]: what's "acfg"?

**Review by robert-zaremba — COMMENTED** [review]
+ in some places we are using `tx` in other `txn`. Let's stick to one - for functions, variables, infixes... If Algorand uses `txn` then maybe we should follow that? On the other hand, in our code we have `tx`. So let's use what makes most of the sense (I think `tx`)

File: packages/runtime/src/ctx.ts [hunk 3]
```diff
@@ -49,6 +49,12 @@ export class Ctx implements Context {
     this.createdAssetID = 0;
   }
 
+  private setAndGetTxnInfo (): TxReceipt {
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/runtime/src/ctx.ts:52]: ok

File: packages/runtime/src/ctx.ts [hunk 4]
```diff
@@ -49,6 +49,12 @@ export class Ctx implements Context {
...
+    const info = { txn: this.tx, txID: this.tx.txID };
+    this.state.txnInfo.set(this.tx.txID, info);
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/runtime/src/ctx.ts:54]: done

File: packages/runtime/src/runtime.ts (hunk 2 above)
Comment (ratik21) [reply packages/runtime/src/runtime.ts:351]: It's a type of transaction (asset deploy/modify/delete ..): [https://developer.algorand.org/docs/get-details/dapps/avm/teal/specification/#typeenum-constants](https://developer.algorand.org/docs/get-details/dapps/avm/teal/specification/#typeenum-constants)

I think using an Enum here would be more intuitive. Updating

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 5]
```diff
@@ -4056,3 +4058,177 @@ export class ITxna extends Op {
...
+export class Args extends Arg {
+  /**
+   * Gets the argument value from interpreter.args array.
+   * store the value in _arg variable
+   * @param args Expected arguments: none
+   * @param line line number in TEAL file
+   * @param interpreter interpreter object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super([...args, "100"], line, interpreter);
```
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/opcode-list.ts:4171]: why adding "100"?

File: packages/runtime/src/lib/constants.ts [hunk 6]
```diff
@@ -32,6 +32,8 @@ export const ALGORAND_MAX_TX_ACCOUNTS_LEN = 4;
 // the assets and application arrays combined and totaled with the accounts array can not exceed 8
 export const ALGORAND_MAX_TX_ARRAY_LEN = 8;
 export const MAX_INNER_TRANSACTIONS = 16;
+export const ALGORAND_MAX_LOGS_COUNT = 32;
+export const ALGORAND_MAX_LOGS_LENGTH = 1024;
```
Comment (robert-zaremba) [inline packages/runtime/src/lib/constants.ts:36]: let's check if this is defined in JS-SDK as well, if so, we should use SDK directly.

**Review by robert-zaremba — APPROVED** [review]

File: packages/runtime/src/interpreter/interpreter.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ import { Stack } from "../lib/stack";
 import { assertMaxCost, parser } from "../parser/parser";
 import {
   AccountStoreI, ExecutionMode, Operator, SSCAttributesM,
-  StackElem, TEALStack, Txn
+  StackElem, TEALStack, Txn, TxReceipt
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/runtime/src/interpreter/interpreter.ts:13]: alright

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 5 above)
Comment (ratik21) [reply packages/runtime/src/interpreter/opcode-list.ts:4171]: just a mock value, to inherit from super class. In this op the value corresponding to "100" will be fetched from stack

File: packages/runtime/src/lib/constants.ts (hunk 6 above)
Comment (ratik21) [reply packages/runtime/src/lib/constants.ts:36]: unfortunately, this is only exported from `go-algorand`

**Pull request merged**